### ISSUE TITLE
Different behavior for a Security enabled cluster using EQL

### ISSUE THREAD

**Issue by astefan** [elastic/elasticsearch#63561]
Prerequisites:
- the `test` index does exist and has data in it
```
{"index":{}}
{"ulong":"123456", "int": 123, "text": "bla", "@timestamp": "2020-12-08T11:04:05.000Z"}
{"index":{}}
{"ulong":"1234567890", "int": -1123, "text": "bla", "@timestamp": "2020-12-08T11:04:05.000Z"}
{"index":{}}
{"ulong":"1234567890123", "int": 444, "text": "bla", "@timestamp": "2020-12-08T11:04:05.000Z"}
```
- there is no `bla*` named index or alias in the cluster

Additional note: EQL uses `allow_no_indices: false` by default.

For a request like
```
/bla*,test/_eql/search
{
	"timestamp_field": "@timestamp",
	"event_category_field": "text",
	"query": "bla where true"
}
```
there are two different behaviors for a cluster with and without Security enabled in it.

Behavior without Security: data is returned and there is no error.
Behavior with Security:
```
{
    "error": {
        "root_cause": [
            {
                "type": "index_not_found_exception",
                "reason": "no such index [bla*]",
                "index_uuid": "_na_",
                "index": "bla*"
            }
        ],
        "type": "index_not_found_exception",
        "reason": "no such index [bla*]",
        "index_uuid": "_na_",
        "index": "bla*"
    },
    "status": 404
}
```

Tracking down this behavior led me to `AuthorizationService.resolveIndexNames` which calls `IndexAbstractionResolver.resolveIndexAbstractions:102` where the exception is being thrown from.

Just a snippet of the calling stack
```
	IndexAbstractionResolver.resolveIndexAbstractions(Iterable<String>, IndicesOptions, Metadata, Collection<String>, boolean, boolean) line: 102	
	IndicesAndAliasesResolver.resolveIndicesAndAliases(IndicesRequest, Metadata, List<String>) line: 151	
	IndicesAndAliasesResolver.resolve(TransportRequest, Metadata, List<String>) line: 111	
	AuthorizationService.resolveIndexNames(TransportRequest, Metadata, List<String>, ActionListener<ResolvedIndices>) line: 585
```

**Comment by astefan** [comment]
If this is relevant, using `/bla,test/_eql/search` (no wildcard) there is no exception being thrown.

**Comment by ywangd** [comment]
What is the reason for `EqlSearchRequest` to mandate `allow_no_indices=false`? It seems to me that all of its sub-actions actually has the opposite setting, i.e. `allow_no_indices=true`, e.g. [`IndexResolver#resolveAsMergedMapping`](https://github.com/elastic/elasticsearch/blob/7.x/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java#L297-L298) and [`BasicQueryClient#query`](https://github.com/elastic/elasticsearch/blob/7.x/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/BasicQueryClient.java#L57-L58).

`allow_no_indices=false` basically tells the index resolver to [throw error](https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-index.html#multi-index) when a wildcard does not match a valid index. This behaviour is the same regardless of whether security is enabled. However, it worked with security off because this option is never used anywhere to resolve indices (since its sub-actions have `true` value for this option). When security is on, it intercepts any IndicesRequest to verify privileges which requires expanding wildcard. It throws error when it sees that the wildcard does not expand to a valid index and `allow_no_indices` is `false`. 

In summary, my question is: why mandating `allow_no_indices=false` for EqlSearchRequest? If it is not strictly necessary, this issue can be fixed just by dropping this option. Otherwise, we'll need dig deeper to see other alternatives. But before proceeding too far, it would be great to check the simpler option.

**Comment by ywangd** [comment]
> If this is relevant, using /bla,test/_eql/search (no wildcard) there is no exception being thrown.
>
This is expected because the request does not contain any wildcard and `ignore_unavailable` is by default `true`.

**Comment by ywangd** [comment]
Just read the QL team's weekly update which includes some information about why `allow_no_indices=false` is required for `/_eql/search`. IIUC, it needs at one valid index so the query can be validated against its mapping, which in turn is built with the FieldCapbilities API. The following code in `IndexResolver#mergedMappings` throws error if no index is available:

https://github.com/elastic/elasticsearch/blob/284aaa8b22accefdaf2fd4a41f705d85a3b2b0ea/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java#L307-L309

However, when `allow_no_indices=false` and wildcard expansion gets nothing, the index resolution will fail earlier so that above code won't even be executed. It does not fail because the index resolution is done with `allow_no_indices=true`. In my opinion:
1) eql is trying to do something that is already handled in a better place
2) the semantics of `allow_no_indices` is misused

Also, when security is enabled, wildcard will be expanded into conrete index names based on user's privileges. So the following line in EqlSession is not doing what you expected:
https://github.com/elastic/elasticsearch/blob/284aaa8b22accefdaf2fd4a41f705d85a3b2b0ea/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlSession.java#L98

For example, with the following request 
```javascript
GET /te*/_eql/search
{
  "timestamp_field": "@timestamp",
  "event_category_field": "text",
  "query": "bla where true"
}
```
The `indexWildcard` in the above line has value of `test` instead of `te*`. The code is essentially trying to double expand index names, which I believe is the cause of #63529.

In summary, I feel we need review how eql handles index expansion so it leverages and aligns with the existing code as much as possible.

**Comment by astefan** [comment]
@ywangd thank you for your analysis. I missed the subtlety of `allow_no_indices` acting only on wildcarded patterns which explains my confusion about
   > If this is relevant, using /bla,test/_eql/search (no wildcard) there is no exception being thrown.

> This is expected because the request does not contain any wildcard and ignore_unavailable is by default true.

Leaving this aside though and getting back to my original report, the error `index_not_found_exception` in case Security is enabled in the cluster comes even before any EQL code is being executed.
Does this mean that Security tries to expand the wildcard expression based on `allow_no_indices` definition which fails, while in the case of a no Security cluster, there is no attempt of expansion until the code reaches EQL where things are not following `allow_no_indices` behavior (where if one wildcarded pattern matches no indices throw an error): EQL looks at the entire list of indices and wildcarded patterns and if the result of all of this is not a valid index, then we throw an exception. Is my understanding correct?

**Comment by ywangd** [comment]
@astefan Your understanding is correct.
> EQL looks at the entire list of indices and wildcarded patterns and if the result of all of this is not a valid index, then we throw an exception.
>
This behaviour is different from what `allow_no_indices=false` is used for. The `allow_no_indices` option is about expansion of a single wildcard, it does not say anything about the final list (the final result of expanding all index patterns).

It seems to me that EQL does not need mandate `allow_no_indices=false` for `EqlSearchRequest` since it does not rely on the behaviour but rather has its own logic to ensure non-empty list of indices.

The tricky part is that eql tries to report `Unknown index` along with the original index pattern. With security enabled, the indices expansion is done before the request reaches eql. When the request arrives, the original index pattern is no longer available. This is normally not an issue, because we usually want to fail ealier for any index pattern related issue. But eql seems to have a different requirement. We (es-security) talk about storing the original index pattern in a ThreadContext header, but no concrete plan yet. If this requirement makes sense for eql, maybe eql could be a driving use case for it. I can bring this up in our weekly meeting for broader discussions.

**Comment by albertzaharovits** [comment]
I concur with Yang's assessment, and I wish to add some clarifications.

> allow_no_indices acting only on wildcarded patterns 

What I understand you mean by this, is that the `allow_no_indices` indices option concerns the individual wildcard pattern expansion, and not the complete indices expression (i.e. the names and patterns separated by commas). This is true in the broad sense. But note that the index expression is resolved fairly [low-level](https://github.com/elastic/elasticsearch/blob/f491422e1edf79ea08d5657b5035f4a5ee444442/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L905), usually in the transport search action. This doesn't prevent callers to also interpret the index option and act on it, which is what happens with Security (and apparently in EQL too).
But for this to work harmoniously, everyone has to agree on a single behaviour for the option.

> Does this mean that Security tries to expand the wildcard expression based on allow_no_indices definition which fails, while in the case of a no Security cluster, there is no attempt of expansion until the code reaches EQL where things are not following allow_no_indices behavior (where if one wildcarded pattern matches no indices throw an error): EQL looks at the entire list of indices and wildcarded patterns and if the result of all of this is not a valid index, then we throw an exception. Is my understanding correct?

That's correct. 

I _think_ it is better for the eql API to respect the `allow_no_indices` option, in the same way that search does (maybe the internal action behind https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-resolve-index-api.html is useful). If the result of the index expression evaluation returns the empty list or if all the indices do not exist, then I think it is reasonable to treat this as a failure scenario in the eql action itself (eg eql query validation error, empty response or index not found exception), as there is no index expression that says "make sure this expression evaluates to at least one index that exists".

**Comment by astefan** [comment]
@ywangd @albertzaharovits thank you for your input on this issue. Following your comments and further investigation I'm closing this issue and conclude the following:
- [`allow_no_indices` options has been changed to be `true` by default in EQL](https://github.com/elastic/elasticsearch/pull/63573) but it's still possible to have it changed by users
- EQL expectation and `allow_no_indices` behavior are different and, at least for the time being, this will remain like this. We will obey `allow_no_indices` behavior by passing its value to both `_field_caps` and `_search` endpoints but if the outcome of resolving the indices patterns results in a no-index list, then EQL will throw an error. EQL needs at least one valid mapping to be able to evaluate the query that's being passed by users. This is also valid for SQL, but SQL doesn't allow changing `allow_no_indices` setting.
- I will follow-up with another PR to include the original pattern in the error message EQL generates. At the moment `*, -*` (coming from Security resolving an invalid pattern) is included in the message. I understand [the meaning of `*, -*` and the reason for using it](https://github.com/elastic/elasticsearch/issues/63529) (`*` matches all, `-*` matches none which results in an overall "match none" choice which many ES APIs expect from an invalid pattern) and I agree this is the correct behavior from Security point of view.
